File: drills/elixir/solutions/nc77_last_stone_weight__find_max_each_round.exs
defmodule Solution do
  # No ordering kept at all: scan for the heaviest, remove it, scan again. O(n)
  # per round against the sorted version's O(n log n) once -- worse overall, but
  # it makes clear that the only operation the problem needs is "give me the
  # largest", which is exactly the interface a heap provides.
  def last_stone_weight(stones), do: smash(stones)

  defp smash([]), do: 0
  defp smash([only]), do: only

  defp smash(stones) do
    heaviest = Enum.max(stones)
    rest = List.delete(stones, heaviest)
    next = Enum.max(rest)
    remaining = List.delete(rest, next)

    case heaviest - next do
      0 -> smash(remaining)
      remainder -> smash([remainder | remaining])
    end
  end
end

File: drills/go/solutions/nc77_last_stone_weight__find_max_each_round.go
package main

import "sort"

func lastStoneWeight(stones []int) int {
	remaining := append([]int{}, stones...)
	// Sort every round and take the two at the end: O(n log n) per smash,
	// but nothing to get wrong.
	for len(remaining) > 1 {
		sort.Ints(remaining)
		first := remaining[len(remaining)-1]
		second := remaining[len(remaining)-2]
		remaining = remaining[:len(remaining)-2]
		if first != second {
			remaining = append(remaining, first-second)
		}
	}
	if len(remaining) == 0 {
		return 0
	}
	return remaining[0]
}
